many: add passphrase authentication support v2

ZeyadYasser opened the pull request:
This PR is a re-implementation of https://github.com/canonical/snapd/pull/14918 with suggestion from @valentindavid https://github.com/canonical/snapd/pull/14918#pullrequestreview-2558738600 of not having volumes authentication options attached to the bootstrapped containers but rather passed along with the rest of other FDE params. A followup PR could be created to tidy up the passed FDE parameters (e.g. data/save bootstrap containers, primary key, volume auth options).

This PR implements passphrase authentication support during installation. Currently all volumes created during installation are passphrase encrypted, i.e. `ubuntu-data` and `ubuntu-save`.

Note: This is still not exposed through the API until target system (#14943) and entropy checks are added. This is done by not including `passphrase-auth` in the list of supported encryption features (Check SD201).

The TPM connection workaround can be removed once https://github.com/canonical/secboot/pull/360 lands.

bboozzoo commented on boot/seal.go:
```diff
@@ -111,7 +111,7 @@ type sealKeyToModeenvFlags struct {
 // in modeenv.
 // It assumes to be invoked in install mode.
-func sealKeyToModeenvImpl(key, saveKey secboot.BootstrappedContainer, primaryKey []byte, model *asserts.Model, modeenv *Modeenv, flags sealKeyToModeenvFlags) error {
+func sealKeyToModeenvImpl(key, saveKey secboot.BootstrappedContainer, primaryKey []byte, volumesAuth *device.VolumesAuthOptions, model *asserts.Model, modeenv *Modeenv, flags sealKeyToModeenvFlags) error {
```
fwiw, we have 7 arguments at this point
```suggestion
func sealKeyToModeenvImpl(
	key, saveKey secboot.BootstrappedContainer,
	primaryKey []byte,
	volumesAuth *device.VolumesAuthOptions,
	model *asserts.Model,
	modeenv *Modeenv,
	flags sealKeyToModeenvFlags,
) error {
```

bboozzoo commented on boot/seal.go:
```diff
@@ -164,13 +164,13 @@ type SealKeyForBootChainsParams struct {
 }
 
-func sealKeyForBootChainsImpl(method device.SealingMethod, key, saveKey secboot.BootstrappedContainer, primaryKey []byte, params *SealKeyForBootChainsParams) error {
+func sealKeyForBootChainsImpl(method device.SealingMethod, key, saveKey secboot.BootstrappedContainer, primaryKey []byte, volumesAuth *device.VolumesAuthOptions, params *SealKeyForBootChainsParams) error {
```
```suggestion
func sealKeyForBootChainsImpl(
	method device.SealingMethod,
	key, saveKey secboot.BootstrappedContainer,
	primaryKey []byte,
	volumesAuth *device.VolumesAuthOptions,
	params *SealKeyForBootChainsParams,
) error {
```

bboozzoo commented on boot/seal.go:
```diff
@@ -164,13 +164,13 @@ type SealKeyForBootChainsParams struct {
 var SealKeyForBootChains = sealKeyForBootChainsImpl
 
-func sealKeyToModeenvForMethod(method device.SealingMethod, key, saveKey secboot.BootstrappedContainer, primaryKey []byte, model *asserts.Model, modeenv *Modeenv, flags sealKeyToModeenvFlags) error {
+func sealKeyToModeenvForMethod(method device.SealingMethod, key, saveKey secboot.BootstrappedContainer, primaryKey []byte, volumesAuth *device.VolumesAuthOptions, model *asserts.Model, modeenv *Modeenv, flags sealKeyToModeenvFlags) error {
```
```suggestion
func sealKeyToModeenvForMethod(
	method device.SealingMethod,
	key, saveKey secboot.BootstrappedContainer,
	primaryKey []byte,
	volumesAuth *device.VolumesAuthOptions,
	model *asserts.Model,
	modeenv *Modeenv,
	flags sealKeyToModeenvFlags,
) error {
```

bboozzoo reviewed (approved):
LGTM

pedronis reviewed (approved):
thank you

ZeyadYasser commented:
@bboozzoo I had to revert the suggestion from https://github.com/canonical/snapd/pull/14918#discussion_r1922168322 because internally secboot actually closes the passed TPM connection, and since keys for both data and save partitions are created we need a new TPM connection for each. My bad for not remembering this when addressing the comment.

ZeyadYasser commented:
Rebased on top of latest fde branch, and updated secboot to include the double TPM connection fix

ZeyadYasser commented:
Removed blocked label since now fde branch is in a more stable state

ZeyadYasser commented:
added blocked label again, as agreement was to merge into master directly
